[CALCITE-4640] Propagate table scan hints to JDBC

kramerul opened the pull request:
We would like to use table scan hints to pass [parameters and variables to HANA views](https://help.sap.com/viewer/88fe5f56472e40cca6ef3c3dcab4855b/2.0.04/en-US/fafb3ea432e54fca9eff11648df5bccd.html).

It should be possible to convert the following Calcite SQL 

```SQL
SELECT * FROM VIEW /*+ PLACEHOLDERS("$$PARAMETER_1$$"='Test') */
```

using a special `SqlDialect` to a HANA specific statement

```SQL
SELECT * FROM VIEW ('PLACEHOLDER' = ('$$PARAMETER_1$$', 'Test'))
```

See also my [mail in the mailing list](https://mail-archives.apache.org/mod_mbox/calcite-dev/202105.mbox/%3CD161B82F-5DEC-49A2-A873-4817F4DEB15F%40contoso.com%3E).

Seel also [CALCITE-4640](https://issues.apache.org/jira/browse/CALCITE-4640)

danny0405 commented on core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java:
```diff
@@ -525,7 +530,38 @@ private static SqlNode groupItem(List<SqlNode> groupKeys,
+    if (e instanceof Hintable) {
+      final ImmutableList<RelHint> hints = e.getHints();
+      if (hints != null && ! hints.isEmpty()) {
+        SqlParserPos pos = identifier.getParserPosition();
```
Does the `Hintable.getHints` has any chance to return null ?

kramerul replied:
No this seems not to be the case. I will remove this check.

danny0405 commented on core/src/test/java/org/apache/calcite/test/JdbcAdapterTest.java:
```diff
@@ -542,6 +550,31 @@
+              return builder
+                  .scan("foodmart", "expense_fact")
+                  .hints(RelHint.builder("PLACEHOLDERS")
+                      .hintOptions(ImmutableMap.of("a", "b"))
```
You can use this method instead `hintOption(String optionKey, String optionValue)`.

kramerul replied:
I will change it accordingly.

danny0405 commented:
Thanks for the contribution @kramerul , the code looks good, only some small suggestions, did HANA use this feature internally now ? Does it work as expected ?

kramerul commented:
Hi @danny0405,
thanks for your review. This feature will not be used inside HANA. We are building a project, which reads analytical data from several sources using calcite. One of those sources is SAP HANA. Calcite is ideal for this type of application. It works really fine.

danny0405 commented:
> Hi @danny0405,
> thanks for your review. This feature will not be used inside HANA. We are building a project, which reads analytical data from several sources using calcite. One of those sources is SAP HANA. Calcite is ideal for this type of application. It works really fine.

There is a test error, seems caused by the `instanceOf` check, can you take a look ?

```java
[BadInstanceof] `e` is an instance of TableScan which is a subtype of Hintable, so this is equivalent to a null check.
    if (e instanceof Hintable) {
          ^
    (see https://errorprone.info/bugpattern/BadInstanceof)
  Did you mean 'if (e != null) {'?
```

kramerul commented:
I remove the check `(e instanceof Hintable)` since it was not required.

danny0405 commented:
The commit message and title should in the format `[CALCITE-{JIRA_ID}] The actual message` , can you log an issue and bind this PR with that ? Also fix the commit message then.

kramerul commented:
I will create an issue and bind it to the PR.

kramerul commented:
I created an issue and also changed the commit message. I'm not sure how to bind this PR to the issue. In JIRA, I added a comment. In this PR, I changed the description accordingly.

kramerul commented:
I added [the proposals](https://issues.apache.org/jira/browse/CALCITE-4640?focusedCommentId=17358773&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-17358773) of @julianhyde

danny0405 reviewed (approved):
+1
